proposals/unsafe-evolution.md

```diff
@@ -0,0 +1,336 @@
+are where memory unsafety occurs, and are responsible for the bulk of CVEs categorized as memory safety issues.
+
+```cs
+void M()
+{
+    int i = 1;
+    int* ptr = &i; // Not unsafe
+    unsafe
+    {
+        Console.WriteLine(*ptr); // Dereference of memory not managed by the runtime. This is unsafe.
+        ref int intRef = Unsafe.AsRef(ptr); // Conversion of memory not managed by the runtime to a `ref`. This is unsafe.
+    }
+}
+```
+
+## Motivation
+
+Background for this feature can also be found in https://github.com/dotnet/designs/blob/main/proposed/caller-unsafe.md, which tracks the broader ecosystem changes that will be needed as part of this proposal.
+These include BCL updates to properly annotate methods as being unsafe, as well as tooling updates for better understanding of where memory unsafety occurs. For C#
+specifically, we want to make sure that memory unsafety is properly tracked by the language; today, it can be difficult to look at a program holistically and understand all locations where
+memory unsafety occurs. This is because various helpers such as the `System.Runtime.CompilerServices.Unsafe`, `System.Runtime.InteropServices.Marshal`, and others do not express that they
+violate memory safety and need special consideration. Methods that then use these helpers aren't immediately obvious, and when auditing code for memory safety issues (either ahead of time
+when doing review, or when trying to determine the cause of a vulnerability that is being reported) it can be difficult to pinpoint the locations that could be contributing to issues.
+
+Historically, `unsafe` in C# has referred to a specific memory-safety hole: the existence of pointer types. The moment that a pointer type is no longer involved, C# is perfectly happy to let
+memory unsafety lie latent in code. It is this issue that we are looking to address with this evolution of `unsafe` in C# and the .NET ecosystem, labeling areas where memory unsafety could
+potentially occur, making it easier for reviewers and auditors to understand the boundaries of potential memory unsafety in a program. Importantly, this means that we will be _changing_
+the meaning of `unsafe`, not just augmenting it. The existence of a pointer is not itself unsafe; the unsafe action is dereferencing the pointer. This extends further to types themselves;
+types cannot be inherently unsafe. It is only the action of using a type that could be unsafe, not the existence of that type. The new meaning of `unsafe` on a type, and what an unsafe context
+is, is covered in [unsafe contexts](#unsafe-contexts).
+
+In order for this information to flow through the system, we therefore need to have a way to mark methods themselves as `unsafe`. Today, `unsafe` as a method modifier has no external impact,
+it only allows pointers to be used in the signature and body of the member. Going forward, `unsafe` as a modifier will actually publicly change the meaning of the member; it will indicate that
```

Comment:
I assume there is going to be some attribute synthesized on the new unsafe methods?